lib/std/io/buffered_in_stream.zig

```diff
@@ -3,16 +3,46 @@ const io = std.io;
 const assert = std.debug.assert;
 const testing = std.testing;
 
-pub fn BufferedInStream(comptime buffer_size: usize, comptime InStreamType: type) type {
+/// Creates a readable stream which buffers data
+/// It also supports 'un-reading' data, so that it can be read again.
+/// This makes look-ahead style parsing much easier.
+pub fn BufferedInStream(comptime buffer_type: std.fifo.LinearFifoBufferType, comptime InStreamType: type) type {
     return struct {
         unbuffered_in_stream: InStreamType,
-        fifo: FifoType = FifoType.init(),
+        fifo: FifoType,
 
         pub const Error = InStreamType.Error;
         pub const InStream = io.InStream(*Self, Error, read);
 
         const Self = @This();
-        const FifoType = std.fifo.LinearFifo(u8, std.fifo.LinearFifoBufferType{ .Static = buffer_size });
+        const FifoType = std.fifo.LinearFifo(u8, buffer_type);
+
+        pub usingnamespace switch (buffer_type) {
+            .Static => struct {
+                pub fn init(unbuffered_in_stream: InStreamType) Self {
+                    return .{
+                        .unbuffered_in_stream = unbuffered_in_stream,
+                        .fifo = FifoType.init(),
+                    };
+                }
+            },
+            .Slice => struct {
+                pub fn init(unbuffered_in_stream: InStreamType, buf: []u8) Self {
+                    return .{
+                        .unbuffered_in_stream = unbuffered_in_stream,
+                        .fifo = FifoType.init(buf),
+                    };
+                }
+            },
+            .Dynamic => struct {
+                pub fn init(unbuffered_in_stream: InStreamType, allocator: *mem.Allocator) Self {
+                    return .{
+                        .unbuffered_in_stream = unbuffered_in_stream,
+                        .fifo = FifoType.init(allocator),
```

Comment:
This dynamic FIFO needs to be deinitialised at some time. Would it make sense to add a special `BufferedInStream.deinit` function for this?